pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -228,7 +233,9 @@ public CompletableFuture<Message<T>> receiveAsync() {
         } catch (PulsarClientException e) {
             return FutureUtil.failedFuture(e);
         }
-        return internalReceiveAsync();
+        return pauseFuture.thenCompose((v) -> {
+            return internalReceiveAsync();
+        });
```

Comment:
The solution would need to be optimized. Although using a CompletableFuture addresses the functional aspects, it's not optimal at all. For example, in this case, `.thenCompose` will create new objects on every call, also when pausing is disabled.